### PR TITLE
`azurerm_media_streaming_policy` - support for more features

### PR DESCRIPTION
**Pull request by teowa**
Support for:
- `common_encryption_cbcs.clear_key_encryption_configuration`
- `common_encryption_cenc.clear_key_encryption_configuration`
- `common_encryption_cenc.clear_track`
- `common_encryption_cenc.content_key_to_track_mapping`
- `envelope_encryption`

Reference:
- [Streaming Policies](https://learn.microsoft.com/en-us/azure/media-services/latest/stream-streaming-policy-concept)
- [Swagger](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/mediaservices/resource-manager/Microsoft.Media/Metadata/stable/2022-08-01/StreamingPoliciesAndStreamingLocators.json#L826)

Test:
```
make acctests SERVICE="media" TESTARGS="-parallel 20 -test.run=TestAccStreamingPolicy_" TESTTIMEOUT='1440m'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/media -parallel 20 -test.run=TestAccStreamingPolicy_ -timeout 1440m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccStreamingPolicy_basic
=== PAUSE TestAccStreamingPolicy_basic
=== RUN   TestAccStreamingPolicy_clearKeyEncryption
=== PAUSE TestAccStreamingPolicy_clearKeyEncryption
=== RUN   TestAccStreamingPolicy_requiresImport
=== PAUSE TestAccStreamingPolicy_requiresImport
=== RUN   TestAccStreamingPolicy_complete
=== PAUSE TestAccStreamingPolicy_complete
=== RUN   TestAccStreamingPolicy_update
=== PAUSE TestAccStreamingPolicy_update
=== CONT  TestAccStreamingPolicy_basic
=== CONT  TestAccStreamingPolicy_complete
=== CONT  TestAccStreamingPolicy_requiresImport
=== CONT  TestAccStreamingPolicy_clearKeyEncryption
=== CONT  TestAccStreamingPolicy_update
--- PASS: TestAccStreamingPolicy_basic (133.83s)
--- PASS: TestAccStreamingPolicy_clearKeyEncryption (136.04s)
--- PASS: TestAccStreamingPolicy_requiresImport (137.51s)
--- PASS: TestAccStreamingPolicy_complete (198.46s)
--- PASS: TestAccStreamingPolicy_update (227.08s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/media 228.374s

```

### CODE REVIEW

File: website/docs/r/media_streaming_policy.html.markdown [hunk 1]
```diff
@@ -88,12 +120,16 @@ The following arguments are supported:
...
 
+* `envelope_encryption` - (Optional) A `envelope_encryption` block as defined below. Changing this forces a new Streaming Policy to be created. 
+
 * `no_encryption_enabled_protocols` - (Optional) A `no_encryption_enabled_protocols` block as defined below. Changing this forces a new Streaming Policy to be created.
 
 ---
 
 A `common_encryption_cbcs` block supports the following:
 
+* `clear_key_encryption_configuration` - (Optional) A `clear_key_encryption_configuration` block defined as below. Changing this forces a new Streaming Policy to be created.
```
Comment (stephybun) [inline website/docs/r/media_streaming_policy.html.markdown:131]:
```suggestion
* `clear_key_encryption_configuration` - (Optional) A `clear_key_encryption_configuration` block as defined below. Changing this forces a new Streaming Policy to be created.
```

File: website/docs/r/media_streaming_policy.html.markdown [hunk 2]
```diff
@@ -104,6 +140,12 @@ A `common_encryption_cbcs` block supports the following:
 
 A `common_encryption_cenc` block supports the following:
 
+* `clear_key_encryption_configuration` - (Optional) A `clear_key_encryption_configuration` block defined as below. Changing this forces a new Streaming Policy to be created.
+
+* `clear_track` - (Optional) One or more `clear_track` blocks defined as below. Changing this forces a new Streaming Policy to be created.
+
+* `content_key_to_track_mapping` - (Optional) One or more `content_key_to_track_mapping` blocks defined as below. Changing this forces a new Streaming Policy to be created.
+
```
Comment (stephybun) [inline website/docs/r/media_streaming_policy.html.markdown:148]:
```suggestion
* `clear_key_encryption_configuration` - (Optional) A `clear_key_encryption_configuration` block as defined below. Changing this forces a new Streaming Policy to be created.

* `clear_track` - (Optional) One or more `clear_track` blocks as defined below. Changing this forces a new Streaming Policy to be created.

* `content_key_to_track_mapping` - (Optional) One or more `content_key_to_track_mapping` blocks as defined below. Changing this forces a new Streaming Policy to be created.

```

File: website/docs/r/media_streaming_policy.html.markdown [hunk 3]
```diff
@@ -114,6 +156,40 @@ A `common_encryption_cenc` block supports the following:
 
 ---
 
+A `clear_key_encryption_configuration` block supports the following:
+
+* `custom_keys_acquisition_url_template` - (Required) Template for the URL of the custom service delivering content keys to end user players. Not required when using Azure Media Services for issuing licenses. The template supports replaceable tokens that the service will update at runtime with the value specific to the request. The currently supported token value is `{AlternativeMediaId}`, which is replaced with the value of `StreamingLocatorId.AlternativeMediaId`. Changing this forces a new Streaming Policy to be created.
```
Comment (stephybun) [inline website/docs/r/media_streaming_policy.html.markdown:161]: Looks like we're exposing implementation detail here, we should remove that.
```suggestion
* `custom_keys_acquisition_url_template` - (Required) The URL template for the custom service that delivers content keys to the end user. This is not required when using Azure Media Services for issuing licenses. Changing this forces a new Streaming Policy to be created.
```

File: website/docs/r/media_streaming_policy.html.markdown [hunk 4]
```diff
@@ -114,6 +156,40 @@ A `common_encryption_cenc` block supports the following:
...
+
+-> **Note** Either `clear_key_encryption_configuration` or `drm` must be specified.
+
+---
+
+A `clear_track` block supports the following:
+
+* `condition` - (Required) One or more `condition` blocks defined as below. Changing this forces a new Streaming Policy to be created.
```
Comment (stephybun) [inline website/docs/r/media_streaming_policy.html.markdown:169]:
```suggestion
* `condition` - (Required) One or more `condition` blocks as defined below. Changing this forces a new Streaming Policy to be created. 
```

File: website/docs/r/media_streaming_policy.html.markdown [hunk 5]
```diff
@@ -114,6 +156,40 @@ A `common_encryption_cenc` block supports the following:
...
+
+---
+
+A `condition` block supports the following:
+
+* `operation` - (Required) Track property condition operation. Possible value is `Equal`. Changing this forces a new Streaming Policy to be created.
+
+* `property` - (Required) Track property type. Possible value is `FourCC`. Changing this forces a new Streaming Policy to be created.
+
+* `value` - (Required) Track property value. Changing this forces a new Streaming Policy to be created.
```
Comment (stephybun) [inline website/docs/r/media_streaming_policy.html.markdown:179]:
```suggestion
* `operation` - (Required) The track property condition operation. Possible value is `Equal`. Changing this forces a new Streaming Policy to be created.

* `property` - (Required) The track property type. Possible value is `FourCC`. Changing this forces a new Streaming Policy to be created.

* `value` - (Required) The track property value. Changing this forces a new Streaming Policy to be created.
```

File: website/docs/r/media_streaming_policy.html.markdown [hunk 6]
```diff
@@ -114,6 +156,40 @@ A `common_encryption_cenc` block supports the following:
...
+
+---
+
+A `content_key_to_track_mapping` block supports the following:
+
+* `label` - (Optional) Label can be used to specify Content Key when creating a Streaming Locator. Changing this forces a new Streaming Policy to be created.
```
Comment (stephybun) [inline website/docs/r/media_streaming_policy.html.markdown:185]:
```suggestion
* `label` - (Optional) Specifies the content key when creating a Streaming Locator. Changing this forces a new Streaming Policy to be created.
```

File: website/docs/r/media_streaming_policy.html.markdown [hunk 7]
```diff
@@ -114,6 +156,40 @@ A `common_encryption_cenc` block supports the following:
...
+
+* `policy_name` - (Optional) Policy used by Default Key. Changing this forces a new Streaming Policy to be created.
+
+* `track` - (Optional) One or more `track` blocks defined as below. Changing this forces a new Streaming Policy to be created.
```
Comment (stephybun) [inline website/docs/r/media_streaming_policy.html.markdown:189]:
```suggestion
* `policy_name` - (Optional) The policy used by the default key. Changing this forces a new Streaming Policy to be created.

* `track` - (Optional) One or more `track` blocks as defined below. Changing this forces a new Streaming Policy to be created.
```

File: website/docs/r/media_streaming_policy.html.markdown [hunk 8]
```diff
@@ -150,6 +226,16 @@ A `enabled_protocols` block supports the following:
 
 ---
 
+A `envelope_encryption` block supports the following:
+
+* `custom_key_acquisition_url_template` - (Optional) Template for the URL of the custom service delivering keys to end user players. Not required when using Azure Media Services for issuing keys. The template supports replaceable tokens that the service will update at runtime with the value specific to the request. The currently supported token values are `{AlternativeMediaId}`, which is replaced with the value of `StreamingLocatorId.AlternativeMediaId`, and `{ContentKeyId}`, which is replaced with the value of identifier of the key being requested. Changing this forces a new Streaming Policy to be created.
```
Comment (stephybun) [inline website/docs/r/media_streaming_policy.html.markdown:231]: Same as above
```suggestion
* `custom_keys_acquisition_url_template` - (Required) The URL template for the custom service that delivers content keys to the end user. This is not required when using Azure Media Services for issuing licenses. Changing this forces a new Streaming Policy to be created.
```

File: website/docs/r/media_streaming_policy.html.markdown [hunk 9]
```diff
@@ -160,6 +246,13 @@ A `no_encryption_enabled_protocols` block supports the following:
 
 * `smooth_streaming` - (Optional) Enable SmoothStreaming protocol or not. Changing this forces a new Streaming Policy to be created.
 
+---
+
+A `track` block supports the following:
+
+* `condition` - (Required) One or more `condition` blocks defined as below. Changing this forces a new Streaming Policy to be created.
```
Comment (stephybun) [inline website/docs/r/media_streaming_policy.html.markdown:253]:
```suggestion
* `condition` - (Required) One or more `condition` blocks as defined below. Changing this forces a new Streaming Policy to be created. 
```

File: internal/services/media/media_streaming_policy_resource.go [hunk 10]
```diff
@@ -117,13 +255,32 @@ func resourceMediaStreamingPolicy() *pluginsdk.Resource {
 				MaxItems: 1,
 				Elem: &pluginsdk.Resource{
 					Schema: map[string]*pluginsdk.Schema{
+						"clear_key_encryption_configuration": {
```
Comment (stephybun) [inline internal/services/media/media_streaming_policy_resource.go:258]: Configuration is implied, we should remove that from the name
```suggestion
						"clear_key_encryption": {
```

File: internal/services/media/media_streaming_policy_resource.go [hunk 11]
```diff
@@ -70,20 +71,91 @@ func resourceMediaStreamingPolicy() *pluginsdk.Resource {
...
```
Comment (stephybun) [inline internal/services/media/media_streaming_policy_resource.go:74]: Same as below
```suggestion
						"clear_key_encryption": {
```

**Review by stephybun — CHANGES_REQUESTED** [review]
Thanks for this @teowa, the docs needs a few adjustments as well as the schema. Otherwise this is looking good.

**Comment by teowa** [comment]
Hi @stephybun , thanks for reviewing.
I have changed the code per review comment, please kindly have another look.

```
TF_ACC=1 go test -v ./internal/services/media -parallel 20 -test.run=TestAccStreamingPolicy_ -timeout 1440m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccStreamingPolicy_basic
=== PAUSE TestAccStreamingPolicy_basic
=== RUN   TestAccStreamingPolicy_clearKeyEncryption
=== PAUSE TestAccStreamingPolicy_clearKeyEncryption
=== RUN   TestAccStreamingPolicy_requiresImport
=== PAUSE TestAccStreamingPolicy_requiresImport
=== RUN   TestAccStreamingPolicy_complete
=== PAUSE TestAccStreamingPolicy_complete
=== RUN   TestAccStreamingPolicy_update
=== PAUSE TestAccStreamingPolicy_update
=== CONT  TestAccStreamingPolicy_basic
=== CONT  TestAccStreamingPolicy_complete
=== CONT  TestAccStreamingPolicy_requiresImport
=== CONT  TestAccStreamingPolicy_clearKeyEncryption
=== CONT  TestAccStreamingPolicy_update
--- PASS: TestAccStreamingPolicy_basic (131.54s)
--- PASS: TestAccStreamingPolicy_complete (131.57s)
--- PASS: TestAccStreamingPolicy_requiresImport (132.65s)
--- PASS: TestAccStreamingPolicy_clearKeyEncryption (193.25s)
--- PASS: TestAccStreamingPolicy_update (208.67s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/media 210.662s

```

**Review by stephybun — APPROVED** [review]
Thanks @teowa LGTM 🦢

**Pull request merged**